src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs

```diff
@@ -0,0 +1,23 @@
+﻿using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Design;
+using Microsoft.EntityFrameworkCore.Scaffolding;
+using Microsoft.EntityFrameworkCore.SqlServer.Storage;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.Design
+{
+    [UsedImplicitly]
+    public class SqlServerNodaTimeDesignTimeServices : IDesignTimeServices
+    {
+        public void ConfigureDesignTimeServices(IServiceCollection serviceCollection)
+            => serviceCollection
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, InstantTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalTimeTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, DurationTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTimeTypeMappingSourcePlugin>()
```

Comment:
It seems odd that there are two plugins that map to datetime2 ? Sould it not just be the Instant mapping only?

Reply:
There are several NodaTime types that while distinct in their concepts in the CLR world, translate to the same MS SQL Server DB type.  As you noted, both the `Instant` type and the `LocalDateTime` type are mapped to the `datetime2` type in the DB.  This is because both NodaTime types require the storage of date and time information in the DB.  I suppose one could map to `datetime` and the other to `datetime2` for purposes of separating the 2 but that would be arbitrary and not really represent the actual difference between the 2 types in the CLR world (and making one of them less precise than the other for no reason).  There are also currently 2 types that map to the `time` SQL type: `LocalTime` and `Duration`.  You can read up more on the NodaTime types [here](https://nodatime.org/3.0.x/userguide/core-types).

That being said, I'm not entirely sure how that would affect the scaffolding process.  We would have to default to one or the other I suppose... and that's probably where the `SqlServerTypeMappingSource` comes into play.  I can look at what the Npgsql provider does and default to the same if it makes sense.  Although IIRC PostgreSQL natively supports all of the NodaTime types and so there must be a way in the DB itself to distinguish between all the different NodaTime types.  Unfortunately that's not the case with SQL Server.

Reply:
PostgreSQL definitely does not have date/time store types which happen to perfectly correspond to the NodaTime CLR types. If you look at [NpgsqlNodaTimeTypeMappingSourcePlugin](https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG.NodaTime/Storage/Internal/NpgsqlNodaTimeTypeMappingSourcePlugin.cs#L54), you'll see that both Instant and LocalDateTime map to PostgreSQL `timestamp without time zone`.

However, it's perfectly normal for more than one CLR type to be mapped to the same database type. When scaffolding, the type mapping source (or plugin) is asked to find a mapping given a store type only (string) and no CLR type, and must return a single mapping (and so a single CLR type). For example, while in Npgsql both Instant and LocalDateTime map to PostgreSQL `timestamp without time zone`, if you scaffold a column of that type you'll get an Instant as the default corresponding type.